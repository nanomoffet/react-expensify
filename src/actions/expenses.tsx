// ADD_EXPENSE

import uuid from 'uuid';

export const addExpense = ({
    description = '',
    note = '',
    amount = 0,
    createdAt = 0,
} = {}) => ({
    type: 'ADD_EXPENSE',
    expense: {
        id: uuid(),
        description,
        note,
        amount,
        createdAt,
    },
});

// REMOVE_EXPENSE

export const removeExpense = ({ id }: { id: string }) => ({
    type: 'REMOVE_EXPENSE',
    id,
});
// EDIT_EXPENSE

export const editExpense = (id: string, updates: any) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates,
});
