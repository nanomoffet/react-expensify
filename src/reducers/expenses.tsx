// Expenses Reducer

const expensesReducerDefaultState: any[] = [];

const expensesReducer = (state = expensesReducerDefaultState, action: any) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return [...state, action.expense];
        case 'REMOVE_EXPENSE':
            return state.filter(({ id }) => id !== action.id);
        case 'EDIT_EXPENSE':
            return state.map((expense: any) => {
                if (expense.id === action.id) {
                    return {
                        ...expense,
                        ...action.updates,
                    };
                } else {
                    return expense;
                }
            });
        default:
            return state;
    }
};

export default expensesReducer;
