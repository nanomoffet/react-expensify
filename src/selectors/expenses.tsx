// Get visible expenses
import moment from 'moment';

const getVisibleExpenses = (
    expenses: any[],
    {
        text,
        sortBy,
        startDate,
        endDate,
    }: { text: any; sortBy: any; startDate: any; endDate: any }
) => {
    return expenses
        .filter(expense => {
            const createdAtMoment = moment(expense.createdAt);
            const startDateMatch = startDate
                ? startDate.isSameOrBefore(createdAtMoment, 'day')
                : true;
            const endDateMatch = endDate
                ? endDate.isSameOrAfter(createdAtMoment, 'day')
                : true;
            const textMatch = expense.description
                .toLowerCase()
                .includes(text.toLowerCase());

            return startDateMatch && endDateMatch && textMatch;
        })
        .sort((a: any, b: any) => {
            if (sortBy === 'date') {
                return a.createdAt < b.createdAt ? 1 : -1;
            } else {
                return a.amount < b.amount ? 1 : -1;
            }
        });
};

export default getVisibleExpenses;
