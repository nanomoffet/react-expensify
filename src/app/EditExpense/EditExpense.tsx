import React, { Component } from 'react';
import { connect } from 'react-redux';
import ExpenseForm from '../AddExpense/ExpenseForm';
import { editExpense, removeExpense } from '../../actions/expenses';

interface props {
    editExpense: any;
    removeExpense: any;
    history: any;
    expense: any;
}

export class EditExpensePage extends Component<props> {
    onSubmit = (expense: any) => {
        this.props.editExpense(this.props.expense.id, expense);
        this.props.history.push('/');
    };

    onRemove = () => {
        this.props.removeExpense({ id: this.props.expense.id });
        this.props.history.push('/');
    };

    render() {
        return (
            <div>
                <button onClick={this.onRemove}>Remove</button>
                <ExpenseForm
                    expense={this.props.expense}
                    onSubmit={this.onSubmit}
                />
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch: any) => ({
    editExpense: (id: any, expense: any) => dispatch(editExpense(id, expense)),
    removeExpense: (expense: any) => dispatch(removeExpense(expense)),
});

const mapStateToProps = (state: any, props: any) => {
    return {
        expense: state.expenses.find(
            (expense: any) => expense.id === props.match.params.id
        ),
    };
};

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(EditExpensePage);
