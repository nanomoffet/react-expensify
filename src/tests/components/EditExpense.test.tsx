import React from 'react';
import { EditExpensePage } from '../../app/EditExpense/EditExpense';
import { shallow } from 'enzyme';
import expenses from '../fixtures/expenses.fixtures';

let editExpense: any, removeExpense: any, history: any, wrapper: any;

beforeEach(() => {
    editExpense = jest.fn();
    removeExpense = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(
        <EditExpensePage
            history={history}
            editExpense={editExpense}
            removeExpense={removeExpense}
            expense={expenses[2]}
        />
    );
});

it('should render EditExpensePage correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

it('should handle editExpense', () => {
    const submit: any = wrapper.find('ExpenseForm').prop('onSubmit');
    submit(expenses[2]);

    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(editExpense).toHaveBeenLastCalledWith(expenses[2].id, expenses[2]);
});

it('should handle removeExpense', () => {
    wrapper.find('button').simulate('click');

    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(removeExpense).toHaveBeenLastCalledWith({
        id: expenses[2].id,
    });
});
