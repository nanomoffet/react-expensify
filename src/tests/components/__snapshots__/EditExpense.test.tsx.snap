// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage correctly 1`] = `
<div>
  <button
    onClick={[Function]}
  >
    Remove
  </button>
  <ExpenseForm
    expense={
      Object {
        "amount": 4500,
        "createdAt": 345600000,
        "description": "Credit Card",
        "id": "3",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
</div>
`;
